dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/directory/AbstractDirectory.java

```diff
@@ -72,8 +75,13 @@ public AbstractDirectory(URL url, RouterChain<T> routerChain) {
         this.consumedProtocol = this.queryMap.get(PROTOCOL_KEY) == null ? DUBBO : this.queryMap.get(PROTOCOL_KEY);
         this.url = url.removeParameter(REFER_KEY).removeParameter(MONITOR_KEY);
 
-        this.consumerUrl = this.url.setProtocol(consumedProtocol).setPath(path == null ? queryMap.get(INTERFACE_KEY) : path).addParameters(queryMap)
-                .removeParameter(MONITOR_KEY);
+        URL consumerUrlFrom = this.url.setProtocol(consumedProtocol)
+                .setPath(path == null ? queryMap.get(INTERFACE_KEY) : path);
+        if (isUrlFromRegistry) {
+            // reserve parameters if url is already a consumer url
+            consumerUrlFrom = consumerUrlFrom.clearParameters();
```

Comment:
Will clearing all the parameters of the registry url cause the loss of some useful parameters of the consumer url?

Reply:
> Will clearing all the parameters of the registry url cause the loss of some useful parameters of the consumer url?

The paramaters in url which is from `DynamicDirectory`  is the paramaters of registry. Real reference paramaters are bring encoded and set in `refer` paramater which is `queryMap` here.